hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/SchemaReplicator.java

```diff
@@ -0,0 +1,279 @@
+
+        return CompletableFuture.allOf(replications);
+    }
+
+    /**
+     * Marks the replication status of the schema as
+     * {@link SchemaReplicationStatus#PREPARED} if it is not marked yet.
+     *
+     * @param schema to set the status.
+     */
+    public void markSchemaAsPrepared(Schema schema) {
+        long schemaId = schema.getSchemaId();
+        SchemaReplication replication = new SchemaReplication(schema, SchemaReplicationStatus.PREPARED);
+        replications.putIfAbsent(schemaId, replication);
+    }
+
+
+    /**
+     * Marks the replication status of the schema as
+     * {@link SchemaReplicationStatus#REPLICATED}.
+     * <p>
+     * It assumes that there is already a replication registered for
+     * that schema id.
+     *
+     * @param schemaId to set the status.
+     */
+    public void markSchemaAsReplicated(long schemaId) {
+        SchemaReplication existing = replications.get(schemaId);
+        assert existing != null;
+        existing.setStatus(SchemaReplicationStatus.REPLICATED);
+    }
+
+    /**
+     * Returns the replication status for the given schema, if it exists.
+     */
+    public SchemaReplicationStatus getReplicationStatus(Schema schema) {
+        SchemaReplication replication = replications.get(schema.getSchemaId());
+        if (replication == null) {
+            return null;
+        }
+
+        return replication.getStatus();
+    }
+
+    /**
+     * Returns the collection of all replications(prepared&replicated) so far.
+     */
+    public Collection<SchemaReplication> getReplications() {
+        // shallow copy is enough, we won't mutate the replications
+        return new ArrayList<>(replications.values());
+    }
+
+    /**
+     * Sets the status of the replications as they are available in the
+     * {@code replications}.
+     * <p>
+     * It assumes that the {@code this.replications} is an empty map.
+     * <p>
+     * This method might be called on two occasions:
+     * <ul>
+     *     <li>For newly joining members, where {@code this.replications} is
+     *     empty</li>
+     *     <li>For members of the smaller cluster when they join the larger
+     *     cluster, during the split brain healing, where
+     *     {@link #clear()} is called beforehand, which clears the
+     *     {@code this.replications}.</li>
+     * </ul>
+     */
+    public void setReplications(Collection<SchemaReplication> replications) {
+        for (SchemaReplication replication : replications) {
+            long schemaId = replication.getSchema().getSchemaId();
+            this.replications.put(schemaId, replication);
+        }
+    }
+
+    private CompletableFuture<Void> doReplicate(Schema schema) {
+        long schemaId = schema.getSchemaId();
+        return prepareOnCaller(schema)
+                .thenCompose(result -> {
+                    replications.put(schemaId, new SchemaReplication(schema, SchemaReplicationStatus.PREPARED));
```

Comment:
```suggestion
                    replications.putIfAbsent(schemaId, new SchemaReplication(schema, SchemaReplicationStatus.PREPARED));
```
And you should check the return value and should NOT continue if already another thread is doing it.

Reply:
I have updated this one, while working on other improvements